File: ml-service/burn_rate.py
# FORE — ml-service/burn_rate.py
# Owner: TASK-003 (Vishvraj). CONTRACT-003 in docs/CONTRACTS.md.
# Linear regression, honestly labeled as a straight-line trend — no forecasting-accuracy claim.
# Reused by TASK-007's can_i_afford.py: insert the hypothetical expense into the transaction set
# fed here, do not write a second regression path.
#
# POST /burn-rate
# Input:  { transactions: Transaction[] }
# Output: { daily_avg, trend_slope, projected_zero_balance_date, weekly_seasonal, ... }

from datetime import date, timedelta
import math

_CREDIT_CATEGORIES = {"income", "salary", "credit", "refund", "cashback", "opening_balance"}
_PROJECTION_CAP_DAYS = 3650


def _parse_date(value: str) -> date:
    return date.fromisoformat(str(value)[:10])


def _signed_amount(txn: dict) -> float:
    amount = float(txn.get("amount", 0))
    category = str(txn.get("category", "")).strip().lower()
    if category == "transfers":
        return amount
    if category in _CREDIT_CATEGORIES:
        return abs(amount)
    return -abs(amount)


def _weekly_seasonal_factors(dated: list[tuple[date, float]]) -> list[float]:
    dow_spend = [0.0] * 7
    dow_count = [0] * 7
    for txn_date, amt in dated:
        if amt >= 0:
            continue
        dow = txn_date.weekday()
        # Python weekday: Mon=0 … convert to Sun=0
        dow_sun = (dow + 1) % 7
        dow_spend[dow_sun] += -amt
        dow_count[dow_sun] += 1
    daily_avgs = [dow_spend[i] / dow_count[i] if dow_count[i] else 0.0 for i in range(7)]
    overall = sum(daily_avgs) / 7 or 1.0
    return [round(v / overall, 3) if overall > 0 else 1.0 for v in daily_avgs]


def compute_burn_rate(transactions: list) -> dict:
    if not transactions:
        raise ValueError("transactions must be a non-empty array")

    dated: list[tuple[date, float]] = []
    for txn in transactions:
        try:
            dated.append((_parse_date(txn["date"]), _signed_amount(txn)))
        except (KeyError, ValueError, TypeError):
            raise ValueError(f"transaction missing a valid ISO date: {txn!r}")
    dated.sort(key=lambda pair: pair[0])

    first_day, last_day = dated[0][0], dated[-1][0]
    window_days = (last_day - first_day).days + 1

    total_spend = sum(-amt for _, amt in dated if amt < 0)
    daily_avg = total_spend / window_days
    weekly_seasonal = _weekly_seasonal_factors(dated)

    daily_net = [0.0] * window_days
    for txn_date, amt in dated:
        daily_net[(txn_date - first_day).days] += amt
    balances: list[float] = []
    running = 0.0
    for net in daily_net:
        running += net
        balances.append(running)

    n = window_days
    if n < 2:
        slope = 0.0
    else:
        mean_t = (n - 1) / 2
        mean_b = sum(balances) / n
        var_t = sum((t - mean_t) ** 2 for t in range(n))
        cov_tb = sum((t - mean_t) * (balances[t] - mean_b) for t in range(n))
        slope = cov_tb / var_t

    if n < 2:
        fitted_last = balances[-1]
    else:
        fitted_last = mean_b + slope * ((n - 1) - mean_t)

    residual_std = 0.0
    if n >= 3:
        residuals = [balances[t] - (mean_b + slope * (t - mean_t)) for t in range(n)]
        mse = sum(r * r for r in residuals) / (n - 2)
        residual_std = math.sqrt(mse)

    if slope < 0 and fitted_last > 0:
        trend_days = min(int(fitted_last / -slope) + 1, _PROJECTION_CAP_DAYS)
    elif fitted_last <= 0:
        trend_days = 0
    else:
        trend_days = _PROJECTION_CAP_DAYS
    trend_zero = last_day + timedelta(days=trend_days)

    # Primary projected zero = runway if income/credits stopped (matches PAST UI caption).
    last_balance = balances[-1] if balances else 0.0
    if last_balance <= 0:
        runway_days = 0
    elif daily_avg <= 0:
        runway_days = _PROJECTION_CAP_DAYS
    else:
        runway_days = min(int(last_balance / daily_avg), _PROJECTION_CAP_DAYS)
    projected = last_day + timedelta(days=runway_days)

    interval_days = int(round(residual_std / abs(slope))) if slope < 0 and residual_std > 0 else 0
    projected_low = last_day + timedelta(days=max(0, runway_days - interval_days))
    projected_high = last_day + timedelta(days=min(runway_days + interval_days, _PROJECTION_CAP_DAYS))

    return {
        "daily_avg": round(daily_avg, 2),
        "trend_slope": round(slope, 2),
        "projected_zero_balance_date": projected.isoformat(),
        "weekly_seasonal": weekly_seasonal,
        "projected_zero_balance_date_low": projected_low.isoformat(),
        "projected_zero_balance_date_high": projected_high.isoformat(),
        "trend_zero_balance_date": trend_zero.isoformat(),
        "runway_days_if_income_stopped": runway_days,
    }

File: lib/ml/burnRate.ts
import type { Transaction } from "@/types/financialContext";

const CREDIT_CATEGORIES = new Set(["income", "salary", "credit", "refund", "cashback", "opening_balance"]);
const PROJECTION_CAP_DAYS = 3650;

function parseDate(value: string): Date {
  return new Date(value.slice(0, 10) + "T00:00:00Z");
}

function signedAmount(txn: Transaction): number {
  const amount = Number(txn.amount);
  const category = String(txn.category ?? "").trim().toLowerCase();
  if (category === "transfers") return amount;
  if (CREDIT_CATEGORIES.has(category)) return Math.abs(amount);
  return -Math.abs(amount);
}

function addDays(d: Date, days: number): Date {
  const out = new Date(d);
  out.setUTCDate(out.getUTCDate() + days);
  return out;
}

function isoDate(d: Date): string {
  return d.toISOString().slice(0, 10);
}

/** Day-of-week spend multipliers (Sun=0 … Sat=6); 1.0 = average. */
function weeklySeasonalFactors(dated: [Date, number][]): number[] {
  const dowSpend = new Array(7).fill(0);
  const dowCount = new Array(7).fill(0);

  for (const [dt, amt] of dated) {
    if (amt >= 0) continue;
    const dow = dt.getUTCDay();
    dowSpend[dow] += -amt;
    dowCount[dow]++;
  }

  const dailyAvgs = dowSpend.map((s, i) => (dowCount[i] ? s / dowCount[i] : 0));
  const overall = dailyAvgs.reduce((a, b) => a + b, 0) / 7 || 1;
  return dailyAvgs.map((v) => (overall > 0 ? Math.round((v / overall) * 1000) / 1000 : 1));
}

export function computeBurnRate(transactions: Transaction[]) {
  if (!transactions.length) throw new Error("transactions must be a non-empty array");

  const dated = transactions.map((txn) => {
    const dt = parseDate(txn.date);
    if (Number.isNaN(dt.getTime())) throw new Error(`transaction missing a valid ISO date: ${JSON.stringify(txn)}`);
    return [dt, signedAmount(txn)] as const;
  });
  dated.sort((a, b) => a[0].getTime() - b[0].getTime());

  const firstDay = dated[0][0];
  const lastDay = dated[dated.length - 1][0];
  const windowDays = Math.floor((lastDay.getTime() - firstDay.getTime()) / 86400000) + 1;

  const totalSpend = dated.filter(([, amt]) => amt < 0).reduce((s, [, amt]) => s + -amt, 0);
  const dailyAvg = totalSpend / windowDays;
  const weeklySeasonal = weeklySeasonalFactors(dated as [Date, number][]);

  const dailyNet = new Array<number>(windowDays).fill(0);
  for (const [txnDate, amt] of dated) {
    const idx = Math.floor((txnDate.getTime() - firstDay.getTime()) / 86400000);
    dailyNet[idx] += amt;
  }

  const balances: number[] = [];
  let running = 0;
  for (const net of dailyNet) {
    running += net;
    balances.push(running);
  }

  const n = windowDays;
  let slope = 0;
  if (n >= 2) {
    const meanT = (n - 1) / 2;
    const meanB = balances.reduce((a, b) => a + b, 0) / n;
    const varT = Array.from({ length: n }, (_, t) => (t - meanT) ** 2).reduce((a, b) => a + b, 0);
    const covTb = Array.from({ length: n }, (_, t) => (t - meanT) * (balances[t] - meanB)).reduce((a, b) => a + b, 0);
    slope = covTb / varT;
  }

  let fittedLast: number;
  if (n < 2) {
    fittedLast = balances[balances.length - 1];
  } else {
    const meanT = (n - 1) / 2;
    const meanB = balances.reduce((a, b) => a + b, 0) / n;
    fittedLast = meanB + slope * ((n - 1) - meanT);
  }

  // Residual std for prediction interval on zero-crossing day
  let residualStd = 0;
  if (n >= 3) {
    const meanT = (n - 1) / 2;
    const meanB = balances.reduce((a, b) => a + b, 0) / n;
    const residuals = balances.map((b, t) => b - (meanB + slope * (t - meanT)));
    const mse = residuals.reduce((s, r) => s + r * r, 0) / (n - 2);
    residualStd = Math.sqrt(mse);
  }

  // Trend-based zero (OLS) — informational; can be far out if net-accumulating.
  let trendDaysToZero: number;
  if (slope < 0 && fittedLast > 0) {
    trendDaysToZero = Math.min(Math.floor(fittedLast / -slope) + 1, PROJECTION_CAP_DAYS);
  } else if (fittedLast <= 0) {
    trendDaysToZero = 0;
  } else {
    trendDaysToZero = PROJECTION_CAP_DAYS;
  }
  const trendZero = isoDate(addDays(lastDay, trendDaysToZero));

  // Primary "projected zero" = runway if income/credits stopped (matches PAST caption).
  const lastBalance = balances[balances.length - 1] ?? 0;
  let runwayDays: number;
  if (lastBalance <= 0) runwayDays = 0;
  else if (dailyAvg <= 0) runwayDays = PROJECTION_CAP_DAYS;
  else runwayDays = Math.min(Math.floor(lastBalance / dailyAvg), PROJECTION_CAP_DAYS);
  const projected = isoDate(addDays(lastDay, runwayDays));

  const intervalDays = slope < 0 && residualStd > 0 ? Math.round(residualStd / Math.abs(slope)) : 0;
  const projectedLow = isoDate(addDays(lastDay, Math.max(0, runwayDays - intervalDays)));
  const projectedHigh = isoDate(addDays(lastDay, Math.min(runwayDays + intervalDays, PROJECTION_CAP_DAYS)));

  return {
    daily_avg: Math.round(dailyAvg * 100) / 100,
    trend_slope: Math.round(slope * 100) / 100,
    projected_zero_balance_date: projected,
    weekly_seasonal: weeklySeasonal,
    projected_zero_balance_date_low: projectedLow,
    projected_zero_balance_date_high: projectedHigh,
    trend_zero_balance_date: trendZero,
    runway_days_if_income_stopped: runwayDays,
  };
}
